libBuild.sh

```diff
@@ -0,0 +1,198 @@
+EXTENSION_DIST_ZIP=extension.zip
+EXTENSION_DIST_PREVIEW_FILE=preview-extensions-ggqizro707
+
+EXTENSION_VERSION=2.2.0
+
+function list_all_regions {
+    aws ec2 describe-regions \
+      --all-regions \
+      --query "Regions[].{Name:RegionName}" \
+      --output text | sort
+}
+
+function fetch_extension {
+    arch=$1
+
+    url="https://github.com/newrelic/newrelic-lambda-extension/releases/download/v${EXTENSION_VERSION}/newrelic-lambda-extension.${arch}.zip"
+    rm -rf $EXTENSION_DIST_DIR $EXTENSION_DIST_ZIP
+    curl -L $url -o $EXTENSION_DIST_ZIP
+}
+
+function download_extension {
+    fetch_extension $@
+
+    unzip $EXTENSION_DIST_ZIP -d .
+    rm -f $EXTENSION_DIST_ZIP
+}
+
+function layer_name_str() {
+    rt_part="LambdaExtension"
+    arch_part=""
+
+    case $1 in
+    "java8.al2")
+      rt_part="Java8"
+      ;;
+    "java11")
+      rt_part="Java11"
+      ;;
+    "python3.6")
+      rt_part="Python36"
+      ;;
+    "python3.7")
+      rt_part="Python37"
+      ;;
+    "python3.8")
+      rt_part="Python38"
+      ;;
+    "python3.9")
+      rt_part="Python39"
+      ;;
+    "nodejs12.x")
+      rt_part="NodeJS12X"
+      ;;
+    "nodejs14.x")
+      rt_part="NodeJS14X"
+      ;;
+    esac
+
+    case $2 in
+    "arm64")
+      arch_part="ARM64"
+      ;;
+    "x86_64")
+      arch_part=""
+      ;;
+    esac
+
+    echo "NewRelic${rt_part}${arch_part}"
+}
+
+function s3_prefix() {
+    name="nr-extension"
+
+    case $1 in
+    "java8.al2")
+      name="java-8"
+      ;;
+    "java11")
+      name="java-11"
+      ;;
+    "python3.6")
+      name="nr-python3.6"
+      ;;
+    "python3.7")
+      name="nr-python3.7"
+      ;;
+    "python3.8")
+      name="nr-python3.8"
+      ;;
+    "python3.9")
+      name="nr-python3.9"
+      ;;
+    "nodejs12.x")
+      name="nr-nodejs12.x"
+      ;;
+    "nodejs14.x")
+      name="nr-nodejs14.x"
+      ;;
+    esac
+
+    echo $name
+}
+
+function hash_file() {
+    if which md5sum; then
+        md5sum $1 | awk '{ print $1 }'
+    else
+        md5 -q $1
+    fi
+}
+
+function publish_layer {
+    layer_archive=$1
+    region=$2
+    runtime_name=$3
+    arch=$4
+
+    layer_name=$( layer_name_str $runtime_name $arch )}
+
+    hash=$( hash_file $layer_archive | awk '{ print $1 }' )
+
+    bucket_name="nr-layers-${region}"
+    s3_key="$( s3_prefix $runtime_name )/${hash}.${arch}.zip"
+
+    compat_list=( $runtime_name )
+    if [[ $runtime_name == "provided" ]]
+    then compat_list=("provided" "provided.al2" "dotnetcore3.1" "dotnet6")
+    fi
+
+    echo "Uploading ${layer_archive} to s3://${bucket_name}/${s3_key}"
+    aws --region "$region" s3 cp $layer_archive "s3://${bucket_name}/${s3_key}"
+
+    echo "Publishing ${runtime_name} layer to ${region}"
+    layer_version=$(aws lambda publish-layer-version \
+      --layer-name ${layer_name} \
+      --content "S3Bucket=${bucket_name},S3Key=${s3_key}" \
+      --description "New Relic Layer for ${runtime_name} (${arch})" \
```

Comment:
Were we going to put version info in here?

Reply:
That's out of scope for this. The request for this ticket is to do the refactor, so that there's one publish function, rather than many, which helps with agent version tagging, and all sorts of other things.